MiniDB/ID.cs

```diff
@@ -0,0 +1,241 @@
+        /// </summary>
+        /// <param name="obj1">The first ID to compare</param>
+        /// <param name="obj2">the second ID to compare the first against.</param>
+        /// <returns>Return true if one or more of the components are different</returns>
+        public static bool operator !=(ID obj1, ID obj2)
+        {
+            bool result = false;
+            if (ReferenceEquals(obj1, obj2))
+            {
+                result = false;
+            }
+            else if (ReferenceEquals(obj1, null))
+            {
+                result = true;
+            }
+            else if (ReferenceEquals(obj2, null))
+            {
+                result = true;
+            }
+            else
+            {
+                result = obj1.id != obj2.id || obj1.hardwareComponent != obj2.hardwareComponent;
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// If other object is an ID, compare to it; else, false.
+        /// </summary>
+        /// <param name="other">The other object to compare to</param>
+        /// <returns>True if both components are equal</returns>
+        public override bool Equals(object other)
+        {
+            bool result = false;
+            ID check = other as ID;
+            if (check != null)
+            {
+                result = this.id.Equals(check.id) && this.hardwareComponent.Equals(check.hardwareComponent);
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Determine if this ID is equal (completely by value) to the other id.
+        /// </summary>
+        /// <param name="other">The other id to compare to</param>
+        /// <returns>0 if both components are equal, else comparison of the hardware components if they are different, else comparison of the system components</returns>
+        public bool Equals(ID other)
+        {
+            // use default comparison
+            return this.id.Equals(other.id) && this.hardwareComponent.Equals(other.hardwareComponent);
+        }
+
+        /// <summary>
+        /// Set the ID
+        ///     hardware component is determined based on the system,
+        ///     id field is determined from the input
+        /// </summary>
+        /// <param name="input">the integer to use for the system component</param>
+        /// <returns>The new object</returns>
+        public ID Set(int input)
+        {
+            this.hardwareComponent = DBHardwareID.IDValueInt();
+            this.id = input;
+            return this;
+        }
+
+        /// <summary>
+        /// Set the ID
+        ///    hardware component is determined
+        ///    id fields is set to next random (to lower chances of collisions)
+        /// </summary>
+        /// <returns>The new object</returns>
+        public ID Set()
+        {
+            this.hardwareComponent = DBHardwareID.IDValueInt();
+            this.id = random.Next(0, int.MaxValue);
```

Comment:
and ID can't just be a counter because we are trying to prevent collisions with data in databases that we don't know about? Probably worth a comment somewhere.